File: web/src/mail_selection.ts
import {
  selectionScope,
  type SelectionChange,
  type SelectionCommand,
  type SelectionRepository,
  type SelectionScope,
  type SelectionSnapshot,
} from "./selection_types";

type Gesture =
  | { all: true; failed?: boolean }
  | {
      all?: false;
      change: SelectionChange;
      range?: [number, number];
      failed?: boolean;
    };

/** One ordered request, at most 32 gestures and one rendered page. The complete
 * captured membership stays in the repository, never in the UI controller. */
export class MailSelection {
  mode = false;
  snapshot?: SelectionSnapshot;
  error?: string;
  warning?: string;
  anchor?: string;
  private anchorPosition?: number;
  private identity?: string;
  private backend?: string;
  private capturedScope?: string;
  private running = false;
  private disposed = false;
  private gestures: Gesture[] = [];
  private watched = new Set<string>();
  private dirty = new Map<string, number>();
  private observationGeneration = 0;
  private refreshVersion = 0;
  private acknowledgedRefresh = 0;
  private positions = new Map<string, number>();
  private chosen = new Set<string>();
  private aliases = new Map<string, string>();
  constructor(
    private repository: SelectionRepository,
    private scope: () => SelectionScope,
    private currentCount: () => number,
    private changed: () => void,
  ) {}
  private canonical(id: string) {
    return this.aliases.get(id) ?? id;
  }
  get pending() {
    return this.mode && (!this.snapshot || this.gestures.length > 0);
  }
  get observing() {
    return (
      this.mode &&
      (this.dirty.size > 0 || this.refreshVersion !== this.acknowledgedRefresh)
    );
  }
  get ready() {
    return (
      this.mode &&
      !this.pending &&
      !this.observing &&
      !this.error &&
      this.count > 0
    );
  }
  get count() {
    let count = this.snapshot?.selected ?? 0;
    const chosen = new Set(this.chosen);
    for (const gesture of this.gestures) {
      if (gesture.failed) continue;
      if (gesture.all) {
        count = this.currentCount();
        for (const id of [...this.watched, ...this.gestureIds])
          chosen.add(this.canonical(id));
        continue;
      }
      const change = gesture.change;
      if (change.kind === "clear") {
        count = 0;
        chosen.clear();
      } else if (change.kind === "set") {
        if (change.clear_others) {
          count = 0;
          chosen.clear();
        }
        const id = this.canonical(change.id);
        if (change.selected && !chosen.has(id)) {
          count++;
          chosen.add(id);
        } else if (!change.selected && chosen.delete(id)) count--;
      } else if (gesture.range) {
        const [start, end] = gesture.range;
        if (!change.additive) {
          count = end - start + 1;
          chosen.clear();
        }
        for (const [id, position] of this.positions)
          if (position >= start && position <= end) chosen.add(id);
      }
    }
    return Math.max(0, Math.min(Number.MAX_SAFE_INTEGER, count));
  }
  selected(original: string) {
    const id = this.canonical(original);
    let selected = this.chosen.has(id);
    for (const gesture of this.gestures) {
      if (gesture.failed) continue;
      if (gesture.all) {
        selected = true;
        continue;
      }
      const change = gesture.change;
      if (change.kind === "clear") selected = false;
      else if (change.kind === "set") {
        if (change.clear_others) selected = false;
        if (this.canonical(change.id) === id) selected = change.selected;
      } else if (gesture.range) {
        if (!change.additive) selected = false;
        const position = this.positions.get(id),
          [start, end] = gesture.range;
        if (position !== undefined && position >= start && position <= end)
          selected = true;
      } else if (change.kind === "range") {
        if (!change.additive) selected = false;
        if (
          id === this.canonical(change.target) ||
          id === this.canonical(change.anchor)
        )
          selected = true;
      }
    }
    return this.mode && selected;
  }
  setObserved(ids: string[]) {
    if (ids.length > 50) throw Error("Observe one mail page at a time.");
    const next = new Set(ids);
    this.markDirty(ids.filter((id) => !this.watched.has(id)));
    for (const id of this.watched) if (!next.has(id)) this.dirty.delete(id);
    this.watched = next;
    this.prune();
    queueMicrotask(() => void this.pump());
  }
  private markDirty(ids: Iterable<string>) {
    const generation = ++this.observationGeneration;
    for (const id of ids) this.dirty.set(id, generation);
  }
  private get gestureIds() {
    const ids = new Set<string>();
    for (const gesture of this.gestures) {
      if (gesture.all) continue;
      if (gesture.change.kind === "set") ids.add(gesture.change.id);
      else if (gesture.change.kind === "range") {
        ids.add(gesture.change.anchor);
        ids.add(gesture.change.target);
      }
    }
    return ids;
  }
  private prune() {
    const needed = new Set(
      [
        ...this.watched,
        ...this.gestureIds,
        ...(this.anchor ? [this.anchor] : []),
      ].map((id) => this.canonical(id)),
    );
    for (const id of this.positions.keys())
      if (!needed.has(id)) this.positions.delete(id);
    for (const id of this.chosen) if (!needed.has(id)) this.chosen.delete(id);
    for (const [id, target] of this.aliases)
      if (!needed.has(target)) this.aliases.delete(id);
  }
  start() {
    if (this.mode || this.disposed) return;
    this.mode = true;
    this.identity = crypto.randomUUID();
    this.capturedScope = JSON.stringify(selectionScope(this.scope()));
    this.snapshot = undefined;
    this.error = undefined;
    this.markDirty(this.watched);
    this.changed();
    void this.pump();
  }
  /** Called before the workspace emits a state change, including direct filter
   * and account edits. Scope changes discard immediately; pages do not. */
  reconcileScope() {
    if (
      this.mode &&
      this.capturedScope !== JSON.stringify(selectionScope(this.scope()))
    )
      this.done(false);
  }
  done(notify = true) {
    this.mode = false;
    this.identity = this.anchor = this.capturedScope = undefined;
    this.anchorPosition = undefined;
    this.snapshot = undefined;
    this.error = this.warning = undefined;
    this.gestures = [];
    this.chosen.clear();
    this.positions.clear();
    this.aliases.clear();
    if (notify && !this.disposed) this.changed();
    void this.pump();
  }
  all() {
    this.start();
    this.enqueue({ all: true });
  }
  clear() {
    if (this.enqueue({ change: { kind: "clear" } }))
      this.anchor = this.anchorPosition = undefined;
  }
  toggle(id: string, clearOthers = false) {
    const selected = clearOthers || !this.selected(id);
    this.start();
    if (
      this.enqueue({
        change: { kind: "set", id, selected, clear_others: clearOthers },
      })
    ) {
      this.anchor = id;
      this.anchorPosition = this.positions.get(this.canonical(id));
    }
  }
  range(id: string, additive = false) {
    if (!this.mode || !this.anchor) {
      this.toggle(id, !additive);
      return;
    }
    const from =
        this.positions.get(this.canonical(this.anchor)) ?? this.anchorPosition,
      to = this.positions.get(this.canonical(id));
    this.enqueue({
      change: { kind: "range", anchor: this.anchor, target: id, additive },
      range:
        from === undefined || to === undefined
          ? undefined
          : [Math.min(from, to), Math.max(from, to)],
    });
  }
  refresh() {
    if (!this.mode) return;
    this.refreshVersion++;
    this.markDirty(this.watched);
    void this.pump();
  }
  retry() {
    this.error = undefined;
    if (this.gestures.length) this.gestures[0].failed = false;
    this.changed();
    void this.pump();
  }
  private enqueue(gesture: Gesture) {
    if (!this.mode) return false;
    if (this.gestures.length >= 32) {
      this.warning =
        "Selection is catching up. Retry that click after it finishes.";
      this.changed();
      return false;
    }
    this.warning = undefined;
    this.gestures.push(gesture);
    this.changed();
    void this.pump();
    return true;
  }
  private accept(
    result: SelectionSnapshot,
    observed: Map<string, number | undefined>,
  ) {
    this.snapshot = result;
    for (const [old, target] of Object.entries(result.aliases))
      this.aliases.set(old, target);
    const visible = new Set(result.visible);
    for (const [original, generation] of observed) {
      const id = this.canonical(original);
      this.chosen.delete(id);
      this.positions.delete(id);
      if (visible.has(id)) this.chosen.add(id);
      if (result.positions[id] !== undefined)
        this.positions.set(id, result.positions[id]);
      if (this.dirty.get(original) === generation) this.dirty.delete(original);
    }
    if (this.anchor)
      this.anchorPosition =
        this.positions.get(this.canonical(this.anchor)) ?? this.anchorPosition;
    this.prune();
  }
  private async pump() {
    if (this.running) return;
    this.running = true;
    try {
      while (true) {
        if (this.backend && this.backend !== this.identity) {
          try {
            await this.repository.selection({
              kind: "release",
              id: this.backend,
            });
            this.backend = undefined;
          } catch {
            if (!this.disposed) {
              this.error =
                "Could not close the previous selection. Retry to select messages.";
              this.changed();
            }
            break;
          }
          continue;
        }
        if (this.disposed || !this.mode || this.error) break;
        const id = this.identity!,
          gesture = this.snapshot ? this.gestures[0] : undefined;
        if (
          this.snapshot &&
          !gesture &&
          !this.dirty.size &&
          this.acknowledgedRefresh === this.refreshVersion
        )
          break;
        const observed = [
          ...(gesture || !this.snapshot
            ? new Set([...this.gestureIds, ...this.watched])
            : this.dirty.keys()),
        ].slice(0, 50);
        const versions = new Map(
          observed.map((id) => [id, this.dirty.get(id)]),
        );
        const refresh = this.refreshVersion;
        const command: SelectionCommand =
          !this.snapshot || gesture?.all
            ? {
                kind: "capture",
                id,
                revision: this.snapshot ? this.snapshot.revision + 1 : 0,
                scope: this.scope(),
                all: !!gesture?.all,
              }
            : !gesture
              ? { kind: "observe", id }
              : {
                  kind: "change",
                  id,
                  expected: this.snapshot.revision,
                  scope: this.scope(),
                  change: gesture.change,
                };
        // This unpredictable token may have committed even if its reply was lost.
        this.backend = id;
        try {
          let result: SelectionSnapshot;
          try {
            result = (await this.repository.selection(
              command,
              observed,
            )) as SelectionSnapshot;
          } catch (cause) {
            const recovered = (await this.repository.selection(
              { kind: "observe", id },
              observed,
            )) as SelectionSnapshot;
            const committed =
              command.kind === "capture"
                ? command.revision
                : command.kind === "change"
                  ? command.expected + 1
                  : this.snapshot!.revision;
            if (recovered.revision !== committed) throw cause;
            result = recovered;
          }
          if (this.identity !== id || this.disposed) continue;
          if (gesture) {
            this.gestures.shift();
            this.markDirty(
              [...this.watched].filter((id) => !observed.includes(id)),
            );
          }
          this.acknowledgedRefresh = refresh;
          this.accept(result, versions);
          this.changed();
        } catch {
          if (this.identity !== id || this.disposed) continue;
          if (gesture) gesture.failed = true;
          this.error =
            "Could not update the selection. Retry, or choose Done and select again.";
          this.changed();
          break;
        }
      }
    } finally {
      this.running = false;
    }
  }
  dispose() {
    this.disposed = true;
    this.done(false);
  }
}

File: flutter/lib/model/mail_selection.dart
import 'dart:async';
import 'dart:collection';
import '../data/selection.dart';
import 'mail.dart' show newDraftIdentity;

class _Gesture {
  _Gesture(this.change, {this.all = false, this.range});
  final Map<String, Object?> change;
  final bool all;
  final (int, int)? range;
  bool failed = false;
}

/// A bounded gesture queue and rendered-row observations, never a full inbox's
/// selected-ID set. The captured membership belongs to the native repository.
class MailSelection {
  MailSelection({
    required this.repository,
    required this.scope,
    required this.currentCount,
    required this.changed,
  });
  final SelectionRepository repository;
  final Map<String, Object?> Function() scope;
  final int Function() currentCount;
  final void Function() changed;
  bool mode = false, _running = false, _disposed = false;
  String? _id, _backend, anchor, error, warning;
  int? _anchorPosition;
  SelectionSnapshot? snapshot;
  final _gestures = Queue<_Gesture>();
  final _watched = <String, int>{};
  final _dirty = <String, int>{};
  int _observationGeneration = 0;
  final _positions = <String, int>{};
  final _chosen = <String>{};
  final _aliases = <String, String>{};
  String _canonical(String id) => _aliases[id] ?? id;
  bool get pending => mode && (snapshot == null || _gestures.isNotEmpty);
  bool get ready => mode && !pending && error == null && count > 0;
  int get count {
    var result = snapshot?.selected ?? 0;
    final chosen = Set<String>.of(_chosen);
    for (final gesture in _gestures) {
      if (gesture.failed) continue;
      final change = gesture.change;
      if (gesture.all) {
        result = currentCount();
        chosen.addAll({..._watched.keys, ..._gestureIds}.map(_canonical));
      } else if (change['kind'] == 'clear') {
        result = 0;
        chosen.clear();
      } else if (change['kind'] == 'set') {
        if (change['clear_others'] == true) {
          result = 0;
          chosen.clear();
        }
        final id = _canonical(change['id'] as String);
        if (change['selected'] == true) {
          if (chosen.add(id)) result++;
        } else if (chosen.remove(id)) {
          result--;
        }
      } else if (gesture.range case (final int start, final int end)) {
        if (change['additive'] != true) {
          result = end - start + 1;
          chosen.clear();
        }
        for (final entry in _positions.entries) {
          if (entry.value >= start && entry.value <= end) {
            chosen.add(entry.key);
          }
        }
      }
    }
    // A literal bound: dart2js shifts are 32-bit, so `1 << 53` would be 0.
    return result.clamp(0, 9007199254740991);
  }

  bool selected(String id) {
    id = _canonical(id);
    var value = _chosen.contains(id);
    for (final gesture in _gestures) {
      if (gesture.failed) continue;
      final change = gesture.change;
      if (gesture.all) {
        value = true;
      } else if (change['kind'] == 'clear') {
        value = false;
      } else if (change['kind'] == 'set') {
        if (change['clear_others'] == true) value = false;
        if (_canonical(change['id'] as String) == id) {
          value = change['selected'] == true;
        }
      } else if (gesture.range case (final int start, final int end)) {
        if (change['additive'] != true) value = false;
        final position = _positions[id];
        if (position != null && position >= start && position <= end) {
          value = true;
        }
      } else if (change['kind'] == 'range') {
        if (change['additive'] != true) value = false;
        if (id == _canonical(change['target'] as String) ||
            id == _canonical(change['anchor'] as String)) {
          value = true;
        }
      }
    }
    return mode && value;
  }

  void watch(String id) {
    _watched[id] = (_watched[id] ?? 0) + 1;
    _markDirty([id]);
    scheduleMicrotask(_pump);
  }

  void unwatch(String id) {
    final remaining = (_watched[id] ?? 1) - 1;
    if (remaining > 0) {
      _watched[id] = remaining;
    } else {
      _watched.remove(id);
      _dirty.remove(id);
      _prune();
    }
  }

  void _markDirty(Iterable<String> ids) {
    final generation = ++_observationGeneration;
    for (final id in ids) {
      _dirty[id] = generation;
    }
  }

  Set<String> get _gestureIds => {
    for (final gesture in _gestures)
      for (final key in ['id', 'anchor', 'target'])
        if (gesture.change[key] case final String id) id,
  };

  void _prune() {
    final needed = {
      ..._watched.keys.map(_canonical),
      ..._gestureIds.map(_canonical),
      if (anchor != null) _canonical(anchor!),
    };
    _positions.removeWhere((id, _) => !needed.contains(id));
    _chosen.removeWhere((id) => !needed.contains(id));
    _aliases.removeWhere((_, target) => !needed.contains(target));
  }

  void start() {
    if (mode || _disposed) return;
    mode = true;
    _id = newDraftIdentity();
    snapshot = null;
    error = null;
    _markDirty(_watched.keys);
    changed();
    unawaited(_pump());
  }

  void done() {
    mode = false;
    _id = null;
    anchor = null;
    _anchorPosition = null;
    snapshot = null;
    error = null;
    warning = null;
    _gestures.clear();
    _chosen.clear();
    _positions.clear();
    _aliases.clear();
    changed();
    unawaited(_pump());
  }

  void all() {
    start();
    _enqueue(_Gesture({}, all: true));
  }

  void clear() {
    if (_enqueue(_Gesture({'kind': 'clear'}))) {
      anchor = null;
      _anchorPosition = null;
    }
  }

  void toggle(String id, {bool clearOthers = false}) {
    final next = clearOthers || !selected(id);
    start();
    if (_enqueue(
      _Gesture({
        'kind': 'set',
        'id': id,
        'selected': next,
        'clear_others': clearOthers,
      }),
    )) {
      anchor = id;
      _anchorPosition = _positions[_canonical(id)];
    }
  }

  void range(String id, {bool additive = false}) {
    if (!mode || anchor == null) {
      toggle(id, clearOthers: !additive);
      return;
    }
    final from = _positions[_canonical(anchor!)] ?? _anchorPosition;
    final to = _positions[_canonical(id)];
    _enqueue(
      _Gesture(
        {'kind': 'range', 'anchor': anchor, 'target': id, 'additive': additive},
        range: from == null || to == null
            ? null
            : (from < to ? from : to, from > to ? from : to),
      ),
    );
  }

  void refresh() {
    if (!mode) return;
    _markDirty(_watched.keys);
    unawaited(_pump());
  }

  void retry() {
    error = null;
    if (_gestures.isNotEmpty) _gestures.first.failed = false;
    changed();
    unawaited(_pump());
  }

  bool _enqueue(_Gesture gesture) {
    if (!mode) return false;
    if (_gestures.length >= 32) {
      warning = 'Selection is catching up. Retry that click after it finishes.';
      changed();
      return false;
    }
    warning = null;
    _gestures.add(gesture);
    changed();
    unawaited(_pump());
    return true;
  }

  void _accept(SelectionSnapshot result, Map<String, int?> observed) {
    snapshot = result;
    _aliases.addAll(result.aliases);
    for (final id in observed.keys) {
      final target = _canonical(id);
      _chosen.remove(target);
      _positions.remove(target);
      if (result.visible.contains(target)) _chosen.add(target);
      if (result.positions[target] case final position?) {
        _positions[target] = position;
      }
      if (_dirty[id] == observed[id]) _dirty.remove(id);
    }
    if (anchor != null) {
      _anchorPosition = _positions[_canonical(anchor!)] ?? _anchorPosition;
    }
    _prune();
  }

  Future<void> _pump() async {
    if (_running) return;
    _running = true;
    try {
      while (true) {
        if (_backend != null && _backend != _id) {
          final old = _backend!;
          try {
            await repository.selection({'kind': 'release', 'id': old});
            _backend = null;
          } catch (_) {
            if (!_disposed) {
              error =
                  'Could not close the previous selection. Retry to select messages.';
              changed();
            }
            break;
          }
          continue;
        }
        if (_disposed || !mode || error != null) break;
        final id = _id!;
        final gesture = snapshot == null || _gestures.isEmpty
            ? null
            : _gestures.first;
        if (snapshot != null && gesture == null && _dirty.isEmpty) break;
        final observed =
            (gesture != null || snapshot == null
                    ? {..._gestureIds, ..._watched.keys}
                    : _dirty.keys)
                .take(50)
                .toList();
        final observationVersions = {for (final id in observed) id: _dirty[id]};
        final command = snapshot == null || gesture?.all == true
            ? <String, Object?>{
                'kind': 'capture',
                'id': id,
                'revision': snapshot == null ? 0 : snapshot!.revision + 1,
                'scope': scope(),
                'all': gesture?.all == true,
              }
            : gesture == null
            ? <String, Object?>{'kind': 'observe', 'id': id}
            : <String, Object?>{
                'kind': 'change',
                'id': id,
                'expected': snapshot!.revision,
                'scope': scope(),
                'change': gesture.change,
              };
        // The token may have committed even if the bridge loses its reply.
        _backend = id;
        try {
          SelectionSnapshot result;
          try {
            result = SelectionSnapshot(
              await repository.selection(command, observed: observed)
                  as Map<String, dynamic>,
            );
          } catch (_) {
            final recovered = SelectionSnapshot(
              await repository.selection({
                    'kind': 'observe',
                    'id': id,
                  }, observed: observed)
                  as Map<String, dynamic>,
            );
            final committedRevision = command['kind'] == 'capture'
                ? command['revision'] as int
                : command['kind'] == 'change'
                ? (command['expected'] as int) + 1
                : snapshot!.revision;
            // One controller owns this unpredictable token and serializes all
            // writes. Exactly the next revision proves its last request committed.
            if (recovered.revision != committedRevision) rethrow;
            result = recovered;
          }
          if (_id != id || _disposed) continue;
          if (gesture != null) {
            _gestures.removeFirst();
            _markDirty(_watched.keys.where((id) => !observed.contains(id)));
          }
          _accept(result, observationVersions);
          changed();
        } catch (_) {
          if (_id != id || _disposed) continue;
          gesture?.failed = true;
          error =
              'Could not update the selection. Retry, or choose Done and select again.';
          changed();
          break;
        }
      }
    } finally {
      _running = false;
    }
  }

  void dispose() {
    _disposed = true;
    done();
  }
}
